src/base/misc/fatfs.c

```diff
@@ -1585,19 +1588,13 @@ void build_boot_blk(fatfs_t *f, unsigned char *b)
   b[0x44] = DOS_HELPER_INT;
 
   /*
-   * IO.SYS from MS-DOS 7 normally re-uses the boot block's error message. We
-   * could give it a distinct one here (simply have t_o point to it).
-   *
-   * But don't forget, it's not a simple ASCIIZ string!!!
-   *
-   * NOTE: There is a discrepancy between IO.SYS's and the bootblock's
-   * interpretation of this address. IO.SYS expects t_o to be
-   * relative to 0, but the value actually stored is relative to 0x7c00.
-   * (Leading IO.SYS to display no error message.)
-   * I will assume IO.SYS to be correct for now.
+   * IO.SYS from MS-DOS 7 normally re-uses the boot block's error message.
+   * Please note that the address points to four bytes before the string to
+   * be displayed so we pad beforehand. Presumably this data would be
+   * meaningful, but as yet we do not know what it might be used for.
    */
   t_o = 0x48;
-  snprintf((char *)b + t_o, eos - t_o, msg_f, f->dir);
+  snprintf((char *)b + t_o, eos - t_o, "\x04\x03\x02\x01" MSG_F, f->dir);
```

Comment:
Ideally we'd set a proper value if we knew what its purpose was. It's certainly not a magic value, the MS-DOS 7 boot from image had some other value, but not zero. The snprintf is just a convenient way to set it in one line, but yes if a byte were zero it wouldn't work.  Do you have a succinct method of doing the same without the issue with embedded nul? The odd thing is that if it's omitted, more than just 4 bytes are lost off the front of the displayed string.